[BugFix] [Enhancement] Fix nullptr and support Iceberg null padding

Samrose-Ahmed opened the pull request:
PR #48151 introduced a regression, where what would previously return an Error now caused a nullptr dereference and crashed the entire CN. This change removes the crashing and, for Iceberg, also adds support for padding evolved fields with null values, as per Iceberg spec.

## Why I'm doing:

## What I'm doing:

Fixes #issue

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

Smith-Cruise commented:
can you put crash log here?

Smith-Cruise commented on be/src/formats/parquet/null_column_reader.h:
```diff
@@ -0,0 +1,41 @@
+    ~NullColumnReader() override = default;
+
+    Status read_range(const Range<uint64_t>& range, const Filter* filter, ColumnPtr& dst) override {
+        dst->append_nulls(range.span_size());
```
How about change `NullColumnReader` -> `DefaultColumnReader`

And calling `dst->append_default(range.span_size())`

Because we may support none-nullable column in the future, append_default is better.

Samrose-Ahmed replied:
Yes Iceberg 3 [will support default values](https://iceberg.apache.org/spec/#default-values). I see that `append_default` will append nulls if the column is a `NullableColumn`. I guess this should be safe since currently Iceberg will enforce that dst is a `NullableColumn`. ( dont' want to accidentally return e.g. an empty string instead of NULL for optional column).

zombee0 commented on be/src/formats/parquet/group_reader.cpp:
```diff
@@ -274,6 +275,14 @@ Status GroupReader::_create_column_reader(const GroupReaderParam::Column& column
         }
+        if (column_reader == nullptr) {
+            if (column.t_iceberg_schema_field == nullptr) {
+                return Status::InternalError("Invalid file: No valid column reader.");
```
@Samrose-Ahmed maybe it will be better that split the bugfix and enhancement, as i know, we only materialize column that has at least one subfiled in parquet reader, for a padding column, it's padded at scanner level. cc @Smith-Cruise

Samrose-Ahmed replied:
Yes its for struct evolution in Iceberg e.g. you had `event: struct<action: string>` and you evolved schema to `event: struct<action: string, id:string>`, according to Iceberg spec, if you select `event.id` you need to pad the old files that don't have `event.id` column with null.

zombee0 replied:
Yes, for this case, we want to query event.id and the former files don't contains this subfield, i actually think we have covered this case, and treat the column as column that not need materialized in the reader.

Samrose-Ahmed replied:
I debugged and it crashes on one of our Iceberg tables. I added a unit test to replicate.

Samrose-Ahmed commented:
> can you put crash log here?

The crash is at this line when `column_reader == nullptr`: https://github.com/StarRocks/starrocks/blob/b93405bcdb830a2bd68010a65097d75911173ede/be/src/formats/parquet/group_reader.cpp#L282

Log:

```
main DEBUG (build 20cb672)
query_id:cf9af543-4f0e-11ef-ad60-3a56daa8801c, fragment_instance:cf9af543-4f0e-11ef-ad60-3a56daa8801e
Hive file path: /data/t8D0YA/ts_hour=2024-07-31-07/b2fd9c85-2c13-4545-b819-7fe66ca9566e.parquet, partition id: -1, length: 11181, offset: 4
tracker:process consumption: 477005088
tracker:query_pool consumption: 25234016
tracker:query_pool/connector_scan consumption: 117899264
tracker:load consumption: 0
tracker:metadata consumption: 0
tracker:tablet_metadata consumption: 0
tracker:rowset_metadata consumption: 0
tracker:segment_metadata consumption: 0
tracker:column_metadata consumption: 0
tracker:tablet_schema consumption: 0
tracker:segment_zonemap consumption: 0
tracker:short_key_index consumption: 0
tracker:column_zonemap_index consumption: 0
tracker:ordinal_index consumption: 0
tracker:bitmap_index consumption: 0
tracker:bloom_filter_index consumption: 0
tracker:compaction consumption: 0
tracker:schema_change consumption: 0
tracker:column_pool consumption: 0
tracker:page_cache consumption: 0
tracker:jit_cache consumption: 0
tracker:update consumption: 0
tracker:chunk_allocator consumption: 0
tracker:clone consumption: 0
tracker:consistency consumption: 0
tracker:datacache consumption: 3184743
tracker:replication consumption: 0
*** Aborted at 1722411050 (unix time) try "date -d @1722411050" if you are using GNU date ***
I20240731 00:30:50.066682 126608369256128 logconfig.cpp:131] Start to release memory of cache
I20240731 00:30:50.066693 126608369256128 logconfig.cpp:133] Release memory of cache success
I20240731 00:30:50.066959 126608369256128 logconfig.cpp:147] je_mallctl execute purge success
I20240731 00:30:50.069147 126608369256128 logconfig.cpp:155] je_mallctl execute dontdump success
PC: @         0x1163d5cd starrocks::parquet::GroupReader::_create_column_reader(starrocks::parquet::GroupReaderParam::Column const&)
*** SIGSEGV (@0x0) received by PID 47216 (TID 0x73264ee006c0) from PID 0; stack trace: ***
    @     0x732a5e4a1ec3 (/usr/lib/x86_64-linux-gnu/libc.so.6+0xa1ec2)
    @         0x149dfaf9 google::(anonymous namespace)::FailureSignalHandler(int, siginfo_t*, void*)
    @     0x732a5e445320 (/usr/lib/x86_64-linux-gnu/libc.so.6+0x4531f)
    @         0x1163d5cd starrocks::parquet::GroupReader::_create_column_reader(starrocks::parquet::GroupReaderParam::Column const&)
    @         0x1163d287 starrocks::parquet::GroupReader::_init_column_readers()
    @         0x1163ac09 starrocks::parquet::GroupReader::init()
    @         0x115c3610 starrocks::parquet::FileReader::_init_group_readers()
    @         0x115be4a4 starrocks::parquet::FileReader::init(starrocks::HdfsScannerContext*)
    @         0x1131b55a starrocks::HdfsParquetScanner::do_open(starrocks::RuntimeState*)
    @         0x112f5137 starrocks::HdfsScanner::open(starrocks::RuntimeState*)
    @         0x11244574 starrocks::connector::HiveDataSource::_init_scanner(starrocks::RuntimeState*)
    @         0x1123d529 starrocks::connector::HiveDataSource::open(starrocks::RuntimeState*)
    @          0xc90edb0 starrocks::pipeline::ConnectorChunkSource::_open_data_source(starrocks::RuntimeState*, bool*)
    @          0xc90f18f starrocks::pipeline::ConnectorChunkSource::_read_chunk(starrocks::RuntimeState*, std::shared_ptr<starrocks::Chunk>*)
    @          0xd1c0d97 starrocks::pipeline::ChunkSource::buffer_next_batch_chunks_blocking(starrocks::RuntimeState*, unsigned long, starrocks::workgroup::WorkGroup const*)
    @          0xc8e1c18 auto starrocks::pipeline::ScanOperator::_trigger_next_scan(starrocks::RuntimeState*, int)::{lambda(auto:1&)#1}::operator()<starrocks::workgroup::YieldContext>(starrocks::workgroup::YieldContext&) const
    @          0xc8e4dd5 void std::__invoke_impl<void, starrocks::pipeline::ScanOperator::_trigger_next_scan(starrocks::RuntimeState*, int)::{lambda(auto:1&)#1}&, starrocks::workgroup::YieldContext&>(std::__invoke_other, starrocks::pipeline::ScanOperator::_trigger_next_scan(starro.
    @          0xc8e4cae std::enable_if<is_invocable_r_v<void, starrocks::pipeline::ScanOperator::_trigger_next_scan(starrocks::RuntimeState*, int)::{lambda(auto:1&)#1}&, starrocks::workgroup::YieldContext&>, void>::type std::__invoke_r<void, starrocks::pipeline::ScanOperator::_tr.
    @          0xc8e4ae2 std::_Function_handler<void (starrocks::workgroup::YieldContext&), starrocks::pipeline::ScanOperator::_trigger_next_scan(starrocks::RuntimeState*, int)::{lambda(auto:1&)#1}>::_M_invoke(std::_Any_data const&, starrocks::workgroup::YieldContext&)
    @          0xcb6c083 std::function<void (starrocks::workgroup::YieldContext&)>::operator()(starrocks::workgroup::YieldContext&) const
    @          0xcb6b56f starrocks::workgroup::ScanTask::run()
    @          0xcbe1b7f starrocks::workgroup::ScanExecutor::worker_thread()
    @          0xcbe182d starrocks::workgroup::ScanExecutor::initialize(int)::{lambda()#1}::operator()() const
    @          0xcbe2be6 void std::__invoke_impl<void, starrocks::workgroup::ScanExecutor::initialize(int)::{lambda()#1}&>(std::__invoke_other, starrocks::workgroup::ScanExecutor::initialize(int)::{lambda()#1}&)
    @          0xcbe28d2 std::enable_if<is_invocable_r_v<void, starrocks::workgroup::ScanExecutor::initialize(int)::{lambda()#1}&>, void>::type std::__invoke_r<void, starrocks::workgroup::ScanExecutor::initialize(int)::{lambda()#1}&>(starrocks::workgroup::ScanExecutor::initialize(.
    @          0xcbe2578 std::_Function_handler<void (), starrocks::workgroup::ScanExecutor::initialize(int)::{lambda()#1}>::_M_invoke(std::_Any_data const&)
    @          0xb0e9a90 std::function<void ()>::operator()() const
    @          0xb76c6ba starrocks::FunctionRunnable::run()
    @          0xb76afb3 starrocks::ThreadPool::dispatch_thread()
    @          0xb779724 void std::__invoke_impl<void, void (starrocks::ThreadPool::*&)(), starrocks::ThreadPool*&>(std::__invoke_memfun_deref, void (starrocks::ThreadPool::*&)(), starrocks::ThreadPool*&)
    @          0xb778bf1 std::__invoke_result<void (starrocks::ThreadPool::*&)(), starrocks::ThreadPool*&>::type std::__invoke<void (starrocks::ThreadPool::*&)(), starrocks::ThreadPool*&>(void (starrocks::ThreadPool::*&)(), starrocks::ThreadPool*&)
    @          0xb777f55 void std::_Bind<void (starrocks::ThreadPool::*(starrocks::ThreadPool*))()>::__call<void, , 0ul>(std::tuple<>&&, std::_Index_tuple<0ul>)
Segmentation fault (core dumped)
```

Samrose-Ahmed committed:
[BugFix] [Enhancement] Fix nullptr and Iceberg struct null padding

PR #48151 introduced a regression, where what would previously return an
Error now caused a nullptr dereference and crashed the entire CN.
This change fixes to handle the case and return nulls.

Signed-off-by: Samrose Ahmed <samroseahmed@gmail.com>

Smith-Cruise commented on be/src/formats/parquet/column_reader.cpp:
```diff
@@ -188,7 +188,6 @@ void ColumnReader::get_subfield_pos_with_pruned_type(const ParquetField& field,
         if (iceberg_it == subfield_name_2_field_schema.end()) {
             // This suitition should not be happened, means table's struct subfield not existed in iceberg schema
             // Below code is defensive
-            DCHECK(false) << "Struct subfield name: " + format_subfield_name + " not found in iceberg schema.";
```
What's the situation the code will go here?

TIcebergSchemaField is the latest schema, and the column's type is original from the newest schema, so here is unreachable.

Samrose-Ahmed replied:
I think this is trying to check when an Iceberg struct evolution has happened but we are scanning a parquet file from before the evolution where the file doesn't contain the newly added subfield column. In that case, btw the comment that this "shouldn't happen" isn't really correct, it's quite normal that it would happen.

Smith-Cruise replied:
I disagree with you, must be another problem.

Iceberg is snapshot isolation. Even if the schema is changing, we will only read files based on the latest completed snapshot.

Samrose-Ahmed replied:
Correct, of course. That's referring to the iceberg schema, this is resolving a local file's Parquet schema with the current global table iceberg schema.

Samrose-Ahmed replied:
OK on reading the code, it seems yeah this is already checked by `IcebergMetaHelper::prepare_read_columns`, I can remove this line change.

Smith-Cruise commented on be/src/formats/parquet/group_reader.cpp:
```diff
@@ -274,6 +275,14 @@ Status GroupReader::_create_column_reader(const GroupReaderParam::Column& column
             RETURN_IF_ERROR(ColumnReader::create(_column_reader_opts, schema_node, column.slot_type(),
                                                  column.t_iceberg_schema_field, &column_reader));
         }
+        if (column_reader == nullptr) {
```
Can you write a ut to cover new lines here?
I'm just curious about what the situation can reach here.

You can refer ut to `iceberg_schema_evolution_file_reader_test.cpp`

zombee0 commented:
@Samrose-Ahmed i think the root cause is in `IcebergMetaHelper::prepare_read_columns`, we checked the validation with iceberg schema, we should check it with `materialized_columns` schema.
```
if (!_is_valid_type(parquet_field, iceberg_it->second)) {
            continue;
}
```

Smith-Cruise commented:
`IcebergMetaHelper::_is_valid_type()` should check based on `TIcebergSchemaField` and `TypeDescriptor` both.

You can take a look at `bool ParquetMetaHelper::_is_valid_type()`, it's checked by TypeDescriptor.

Samrose-Ahmed commented:
That makes sense actually I'll update the PR

Samrose-Ahmed commented:
I've updated the PR.

Samrose-Ahmed commented:
Actually I'm getting SIGSEGV at format_name now hmm will have to look into this:

```
*** Aborted at 1722834370 (unix time) try "date -d @1722834370" if you are using GNU date ***
PC: @         0x100517ea std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_data() const
*** SIGSEGV (@0x0) received by PID 1140951 (TID 0x7bd65e0006c0) from PID 0; stack trace: ***
    @     0x7bd723ea1ec3 (/usr/lib/x86_64-linux-gnu/libc.so.6+0xa1ec2)
    @         0x19c4e939 google::(anonymous namespace)::FailureSignalHandler(int, siginfo_t*, void*)
    @     0x7bd723e45320 (/usr/lib/x86_64-linux-gnu/libc.so.6+0x4531f)
    @         0x100517ea std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_data() const
    @         0x10059202 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::end() const
    @         0x10065c68 boost::range_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, void>::type boost::range_detail::range_end<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const>(std::__cxx11::bas4
    @         0x10063b2d boost::range_iterator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, void>::type boost::range_adl_barrier::end<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(std::__cxx11::basic_str4
    @         0x10703393 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > boost::algorithm::detail::transform_range_copy<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_trait4
    @         0x10702406 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > boost::algorithm::to_lower_copy<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(std::__cxx11::basic_string<char, std::char_traits<char>, std::4
    @         0x1349d100 starrocks::Utils::format_name(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool)
    @         0x1461fe2d starrocks::parquet::IcebergMetaHelper::_is_valid_type(starrocks::parquet::ParquetField const*, starrocks::TIcebergSchemaField const*, starrocks::TypeDescriptor const*) const
    @         0x146204c1 starrocks::parquet::IcebergMetaHelper::prepare_read_columns(std::vector<starrocks::HdfsScannerContext::ColumnInfo, std::allocator<starrocks::HdfsScannerContext::ColumnInfo> > const&, std::vector<starrocks::parquet::GroupReaderParam::Column, std::allocator<4
    @         0x146129fa starrocks::parquet::FileReader::_prepare_read_columns(std::unordered_set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equa4
```

zombee0 commented on be/src/formats/parquet/group_reader.cpp:
```diff
@@ -274,6 +274,10 @@ Status GroupReader::_create_column_reader(const GroupReaderParam::Column& column
         }
+        if (column_reader == nullptr) {
+            // this shouldn't happen but guard
+            return Status::InternalError("No valid column reader.");
```
good job

zombee0 reviewed (approved):
LGTM, for clang-format, could you run `./clang-format.sh` in `/path/to/starrocks/code/build-support`

Samrose-Ahmed commented:
This approach doesn't seem to completely work.

I updated it to use field ids but now the new test (TestStructEvolutionPadNull) fails at the same original issue (now returns 'No valid column reader', would crash before w/o check). 

The suggested change is fine to handle robustness for type check but it doesn't seem to handle the root issue, which is you need to pad the nested subfield with default value, that doesn't happen with this code (like comment mentions `we put nullptr column reader in children_reader, we will append default value for this subfield later.` but its never appended later).

It seems like we do need the change I made earlier? thoughts @Smith-Cruise @zombee0

zombee0 commented on be/src/formats/parquet/meta_helper.cpp:
```diff
@@ -120,32 +120,57 @@ void IcebergMetaHelper::_init_field_mapping() {
-            continue;
+        std::unordered_map<int32_t, const TypeDescriptor*> field_id_2_type{};
+        for (size_t idx = 0; idx < type_descriptor->children.size(); idx++) {
+            field_id_2_type.emplace(parquet_field->children[idx].field_id, &type_descriptor->children[idx]);
```
oh, i think this code is not correct, because in the test case, col->d is not exist in parquet field, so the field_id_2_type should be empty, but for this code, field_id_2_type will never be empty.

Samrose-Ahmed replied:
Ok i added a fix for that, but this still doesn't solve that the subfield is not being set to null, this just skips the whole column and sets it to null which is incorrect:

```
[ RUN      ] IcebergSchemaEvolutionTest.TestStructEvolutionPadNull
../test/formats/parquet/iceberg_schema_evolution_file_reader_test.cpp:267: Failure
Expected equality of these values:
  "[{d:NULL}]"
  chunk->debug_row(0)
    Which is: "[NULL]"
[  FAILED  ] IcebergSchemaEvolutionTest.TestStructEvolutionPadNull (0 ms)
[----------] 1 test from IcebergSchemaEvolutionTest (0 ms total)
```

Smith-Cruise replied:
I got what you mean, let me check with Trino's behavior.

Smith-Cruise replied:
I think currently implementation is correct.
```sql
mysql> select * from struct_add_subfield;
+------+---------------------------+
| id   | col_struct                |
+------+---------------------------+
|    1 | {"age":15,"name":null}    |
|    2 | {"age":15,"name":"smith"} |
+------+---------------------------+
2 rows in set (0.07 sec)
```

I've built a table, and the subfield `col_struct.name` is newly added, which means `id=1`'s parquet didn't contain `col_struct.name`.

In spark:
```sql
spark-sql (chendingchao_ice)> select col_struct.name from struct_add_subfield;
NULL
smith
```

In trino:
```sql
trino:chendingchao_ice> select col_struct.name from struct_add_subfield;
 name
-------
 NULL
 smith
(2 rows)
```

In sr, your currently implementation:
```sql
mysql> select col_struct.name from struct_add_subfield;
+-------+
| name  |
+-------+
| NULL  |
| smith |
+-------+
2 rows in set (0.06 sec)
```

Samrose-Ahmed replied:
Ok i just checked with Athena and it seems you are right it does this behavior as well, let me update the test.

zombee0 replied:
Good job, and thank you!

Smith-Cruise commented:
branch 3.2 you need backport either, plz check it.
